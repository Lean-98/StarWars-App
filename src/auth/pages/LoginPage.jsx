import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../context";

export const LoginPage = () => {
  const { login } = useContext(AuthContext);
  const navigate = useNavigate();

  const onLogin = () => {
    const lastPath = localStorage.getItem("lastPath");

    login("Leandro Bazan");

    navigate(lastPath, {
      replace: true,
    });
  };

  return (
    <section className="login">
      <div className="card mb-1">
        <div className="row row-cols-lg-2">
          <div className="col col-sm-0 col-md-6 col-lg-7 d-lg-flex mb-3">
            <img
              src="..\assets\characterStarWars\SEOStarWars.webp"
              alt="Star Wars"
              className="img-responsive rounded-t-5 rounded-tr-lg-4 rounded-bl-lg-5"
            />
          </div>
          <div className="col col-sm-12 col-md-6 col-lg-5  text-center text-lg-start">
            <div className="card-body py-5 px-md-5">
              <form>
                <div className="form-outline mb-4">
                  <label className="form-label text-warning">
                    Email address
                  </label>
                  <input
                    type="email"
                    name="email"
                    className="form-control"
                    required
                  />
                </div>

                <div className="form-outline mb-4">
                  <label className="form-label text-warning">Password</label>
                  <input
                    type="password"
                    name="password"
                    className="form-control"
                    required
                  />
                </div>

                <button className="btn btn-warning btn-lg" onClick={onLogin}>
                  Login
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
    </section>
  );
};
