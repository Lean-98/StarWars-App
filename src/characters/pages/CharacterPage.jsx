import { useMemo } from 'react';
import { Navigate, useNavigate, useParams } from 'react-router-dom';
import { getCharacterById } from '../helpers';

export const CharacterPage = () => {
  const { id } = useParams();
  const navigate = useNavigate();

  const character = useMemo(() => getCharacterById(id), [id]);

  const onNavigateBack = () => {
    navigate(-1);
  };

  if (!character) {
    return <Navigate to={"/"} />;
  }

  return (
    <div className="container">
      <div className="row mt-5">
        <div className="col-sm-6 col-md-5 col-lg-5 text-center mb-4">
          <img
            src={`/assets/characterStarWars/${id}.webp`}
            alt={character.character}
            className="img-fluid animate__animated animate__fadeInLeft"
          />
        </div>

        <div className="col-sm-6 col-md-7 col-lg-7 text-center">
          <ul className="list-group list-group-flush">

             <li className={"list-group-item " + ( character.publisher === 'Jedi' ? 'list-group-item-success' : 'list-group-item-danger')}>
              <h3 className='fw-bolder'>{character.character}</h3>
              <b>Affiliation:</b> {character.publisher}{" "}
              <b>Rank:</b> {character.rank}{" "}
              <br />
              <b>Weapon:</b> {character.Weapon}{" "}
            </li>

            <li className="list-group-item">
            <h5 className="fw-bolder mt-3">First Appearance:</h5>
            <p>{character.first_appearance}</p>
            <h5 className="fw-bolder mt-3">Description:</h5>
            <p>{character.description}</p>

          <button className={"btn " + ( character.publisher === 'Jedi' ? 'btn-success' : 'btn-danger')} onClick={onNavigateBack}>
            Return
          </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
  );
};
