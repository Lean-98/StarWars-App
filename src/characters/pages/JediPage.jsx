import { ListCharacterStarWars } from '../components';


export const JediPage = () => {
  return (
    <>
    <h1 className="fw-semibold text-success text-center p-4">The Most Powerful Jedis of The Saga</h1>

    <ListCharacterStarWars publisher='Jedi' />

    </>
  )
}
