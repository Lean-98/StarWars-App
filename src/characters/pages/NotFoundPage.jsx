export const NotFoundPage = () => {
  return (
    <div className="text-center mt-5" id="error-page">
      <h1>Oops!</h1>
      <p>Sorry, an unexpected error has occurred.</p>
    </div>
  );
};
