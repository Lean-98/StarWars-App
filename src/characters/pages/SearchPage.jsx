import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';
import { useForm } from '../../hooks/useForm';
import { CharacterStarWars } from '../components';
import { getCharacterByName } from '../helpers';

export const SearchPage = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { q = "" } = queryString.parse(location.search);
  const characters = getCharacterByName(q);

  const showSearch = q.length === 0;
  const showError = q.length > 0 && characters.length === 0;

  const { searchText, onInputChange } = useForm({
    searchText: q,
  });

  const onSearchSubmit = (e) => {
    e.preventDefault();
    // if( searchText.trim().length <= 1 ) return;

    navigate(`?q=${searchText.toLowerCase().trim()}`);
  };

  return (
    <>
      <div className="container">
        <div className="row row-cols-md-2 row-cols-lg-2 row-cols-xl-2 g-2">
          <h1 className="fw-semibold text-center p-4">Most Powerful Characters of The Star Wars Saga</h1>
          <hr />
          <div className="col-sm-12 col-md-6 col-lg-6 mb-3">
            <h4 className="d-flex justify-content-around justify-content-md-start">
              Searching
            </h4>
            <hr />
            <form onSubmit={onSearchSubmit}>
              <input
                type="text"
                placeholder="Search a character"
                className="form-control"
                name="searchText"
                autoComplete="off"
                value={searchText}
                onChange={onInputChange}
              />

              <button className="btn btn-warning mt-2">Search</button>
            </form>
          </div>

          <div className="col-sm-12 col-md-6 col-lg-6">
            <h4 className="d-flex justify-content-around justify-content-md-start text-center">
              Results
            </h4>
            <hr />

            {/* {
              ( q === '' )
                ? <div className="alert alert-primary">Search a hero</div>
                : ( heroes.length === 0 ) 
                  && <div className="alert alert-danger">No hero with <b>{ q }</b></div>
            } */}

            <div
              className="alert alert-warning animate__animated animate_fadeIn"
              style={{ display: showSearch ? "" : "none" }}
            >
              Search for a character from the star wars saga
            </div>

            <div
              className="alert alert-danger animate__animated animate_fadeIn"
              style={{ display: showError ? "" : "none" }}
            >
              No character with <b>{q}</b>
            </div>

            {characters.map((character) => (
              <CharacterStarWars key={character.id} {...character} />
            ))}
          </div>
        </div>
      </div>
    </>
  );
};
