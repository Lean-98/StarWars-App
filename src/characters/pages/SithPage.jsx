import { ListCharacterStarWars } from '../components';


export const SithPage = () => {
  return (
    <>
    <h1 className="fw-semibold text-danger text-center p-4">The Most Powerful Sith Lords to Ever Exist</h1>
    
    <ListCharacterStarWars publisher='Sith' />
    </>
  )
}
