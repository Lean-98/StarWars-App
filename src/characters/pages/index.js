export * from "./SearchPage";
export * from "./NotFoundPage";
export * from "./JediPage";
export * from "./SithPage";
export * from "./CharacterPage";
