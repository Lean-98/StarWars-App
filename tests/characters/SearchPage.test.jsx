import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { SearchPage } from '../../src/characters/pages/SearchPage';

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate,
}));

describe('Test on <SearchPage />', () => { 

    beforeEach( () => jest.clearAllMocks() );

    test('should be displayed correctly with default values', () => { 

        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );
        expect( container ).toMatchSnapshot();

     });

     test('should show yoda and the input with the value of the queryString', () => { 

        render(
            <MemoryRouter initialEntries={['/search?q=yoda']}>
                <SearchPage />
            </MemoryRouter>
        );

        const inputValue = screen.getByRole('textbox');
        expect( inputValue.value ).toBe('yoda');

        const img = screen.getByRole('img');
        expect( img.src ).toContain('/assets/characterStarWars/jedi-yoda.webp');
        
        const alert = screen.getByLabelText('alert-danger');
        expect( alert.style.display ).toBe('none');
        // screen.debug();
     });

     test('should display an error if the character is not found(yoda123)', () => { 

        render(
            <MemoryRouter initialEntries={['/search?q=yoda123']}>
                <SearchPage />
            </MemoryRouter>
        );
        const alert = screen.getByLabelText('alert-danger');
        expect( alert.style.display ).toBe('');

      });


     test('should call the navigate to the new screen', () => { 

        const inputValue = 'yoda';

        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        fireEvent.change( input, { target: { name: 'searchText', value: inputValue }});
        console.log(input.value);

        const form = screen.getByRole('form');
        fireEvent.submit( form );
        
        expect( mockUseNavigate ).toHaveBeenCalledWith(`?q=${inputValue}`);

      });

 });