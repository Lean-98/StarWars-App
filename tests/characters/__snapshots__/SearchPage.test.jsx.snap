// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <SearchPage /> should be displayed correctly with default values 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row row-cols-md-2 row-cols-lg-2 row-cols-xl-2 g-2"
    >
      <h1
        class="fw-semibold text-center p-4"
      >
        Most Powerful Characters of The Star Wars Saga
      </h1>
      <hr />
      <div
        class="col-sm-12 col-md-6 col-lg-6 mb-3"
      >
        <h4
          class="d-flex justify-content-around justify-content-md-start"
        >
          Searching
        </h4>
        <hr />
        <form
          aria-label="form"
        >
          <input
            autocomplete="off"
            class="form-control"
            name="searchText"
            placeholder="Search a character"
            type="text"
            value=""
          />
          <button
            class="btn btn-warning mt-2"
          >
            Search
          </button>
        </form>
      </div>
      <div
        class="col-sm-12 col-md-6 col-lg-6"
      >
        <h4
          class="d-flex justify-content-around justify-content-md-start text-center"
        >
          Results
        </h4>
        <hr />
        <div
          class="alert alert-warning animate__animated animate_fadeIn"
        >
          Search for a character from the star wars saga
        </div>
        <div
          aria-label="alert-danger"
          class="alert alert-danger animate__animated animate_fadeIn"
          style="display: none;"
        >
          No character with 
          <b />
        </div>
      </div>
    </div>
  </div>
</div>
`;
